internal/extension/reconciler.go

```diff
@@ -4,12 +4,97 @@ import (
 	"context"
 	"sync"
 	"sync/atomic"
+	"time"
 
 	"github.com/kuadrant/policy-machinery/controller"
 	"github.com/kuadrant/policy-machinery/machinery"
 )
 
-var DAG *atomic.Pointer[StateAwareDAG]
+// nilGuardedPointer is an atomic pointer that provides blocking behavior
+// until the pointer is set to a non-nil value.
+type nilGuardedPointer[T any] struct {
+	ptr  atomic.Pointer[T]
+	mu   sync.Mutex
+	cond *sync.Cond
+}
+
+// newNilGuardedPointer creates a new nilGuardedPointer.
+func newNilGuardedPointer[T any]() *nilGuardedPointer[T] {
+	cgp := &nilGuardedPointer[T]{}
+	cgp.cond = sync.NewCond(&cgp.mu)
+	return cgp
+}
+
+// set sets the pointer to x and signals any goroutines waiting for a non-nil value.
+func (c *nilGuardedPointer[T]) set(x T) {
+	c.ptr.Store(&x)
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cond.Broadcast()
+}
+
+// get returns the current value of the pointer without blocking.
+func (c *nilGuardedPointer[T]) get() *T {
+	return c.ptr.Load()
+}
+
+// getWait blocks until the pointer is set to a non-nil value and then returns that value.
+func (c *nilGuardedPointer[T]) getWait() T {
+	// First try a quick non-blocking check
+	if val := c.ptr.Load(); val != nil {
+		return *c.ptr.Load()
+	}
```

Comment:
Could we use the val we got from the first load? just wondering if there's a chance for it to have been changed and it to be nil on second one:
```suggestion
        val := c.ptr.Load()
	if val != nil {
		return *val
	}
```

Reply:
Yeah, we could, but if you look at the `.Set` it takes a value. So it can only ever go from `nil` to actual value, so while the value _could_ change in between the 2 `Load()`s, it merely would be "a newer value"

Reply:
Yep sure I see that now. Woops sorry, I broke the indentation...